### PR TITLE
🌱 test/e2e: deprecate KCPUpgradeSpec, make ClusterUpgradeConformanceSpec more flexible

### LINKED ISSUE #4896: Depreacate partial upgrade E2Etest
**Detailed Description**

Originally in our E2E test we implemented upgrade tests for single CAPI resources:

- KCPUpgradeSpec
- MachineDeploymentUpgradesSpec 
- MachinePoolSpec

This is now superseed by ClusterUpgradeConformanceSpec, which upgrades all the resources, makes sure CP is upgraded before nodes (thus being compliant with K8s recommendations), and runs conformance on the resulting test.

We should now consider if to deprecate or refactor the legacy tests. e.g.
- KCPUpgradeSpec might still have sense, because it tries different topologies of control plane vs what we usually have in KCPUpgradeSpec
- MachineDeploymentUpgradesSpec should be re-focused to test rollout when the infrastructure template changes.
- MachinePoolSpec should be re-focused on test MachinePool machinery (except upgrades)

However, before taking a decision we should investigate if/how those test are used in the providers...
 
/kind cleanup

@sbueringer

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4896

### CODE REVIEW

File: test/e2e/cluster_upgrade_test.go [hunk 1]
```diff
@@ -51,3 +52,51 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass", func()
 		}
 	})
 })
+
+var _ = Describe("When testing KCP upgrade in a single control plane cluster", func() {
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_test.go:56]: I've kept the test names the same for now. Should we rename them in something like:
* `When upgrading a workload cluster with a single control plane machine`
* `When upgrading a workload cluster with a HA control plane`
* `When upgrading a workload cluster with a HA control plane using scale-in rollout`

I think I'm in favor of renaming

File: test/e2e/kcp_upgrade.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ import (
 )
 
 // KCPUpgradeSpecInput is the input for KCPUpgradeSpec.
+// Deprecated: This struct has been deprecated  and will be removed in a future release.
+// Please use ClusterUpgradeConformanceSpec and ClusterUpgradeConformanceSpecInput instead.
 type KCPUpgradeSpecInput struct {
```
Comment (fabriziopandini) [inline test/e2e/kcp_upgrade.go:37]: > The test module, clusterctl, and experiments do not provide any backward compatible guarantees.

Let's remove these methods and add a note in https://cluster-api.sigs.k8s.io/developer/providers/v1.0-to-v1.1.html

File: test/e2e/cluster_upgrade_test.go (hunk 1 above)
Comment (fabriziopandini) [reply test/e2e/cluster_upgrade_test.go:56]: +1 for renaming

File: test/e2e/kcp_upgrade.go (hunk 2 above)
Comment (sbueringer) [reply test/e2e/kcp_upgrade.go:37]: Perfect, will do

**Comment by sbueringer** [comment]
/hold
Let's run the full e2e test once the e2e tests are working again after https://github.com/kubernetes-sigs/cluster-api/pull/5766 has been merged

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Pull request merged**